### PR TITLE
RoboCup Changes: Port IndirectFreeKickPlay

### LINKED ISSUE #893: RoboCup Changse: Add IndirectFreeKickPlay
### Description of the task

The IndirectFreeKickPlay was added at RoboCup to replace the FreeKickPlay.

### Acceptance criteria

- [ ] Add the IndirectFreeKickPlay

### PR DESCRIPTION
**Pull request by MathewMacDougall**
### Description

Ported the new Indirect free kick play from RoboCup 2019

### Testing Done

None. All Plays will be tested when the new simulator is done.

### Resolved Issues

resolves #893 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/ai/hl/stp/play/corner_kick_play.h [hunk 1]
```diff
@@ -22,11 +22,11 @@ class CornerKickPlay : public Play
 
     void getNextTactics(TacticCoroutine::push_type &yield) override;
 
-   private:
     // The maximum distance from the corner that the ball can be for it to be
     // considered a corner kick
     static constexpr double BALL_IN_CORNER_RADIUS = 0.5;
 
+   private:
```
Comment (jonathanlew) [inline src/software/ai/hl/stp/play/corner_kick_play.h:29]: why was this moved?

File: src/software/ai/hl/stp/play/indirect_free_kick_play.cpp [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+    // on the opposite side of the x-axis to wherever the pass is coming from
+    if (world.ball().position().x() > -1)
+    {
+        double y_offset =
+            -std::copysign(world.field().xLength() / 2, world.ball().position().y());
+        cherry_pick_1_target_region =
+            Rectangle(Point(0, world.field().xLength() / 4),
+                      Point(world.field().xLength() / 2, y_offset));
+        cherry_pick_2_target_region =
+            Rectangle(Point(0, world.field().xLength() / 4), Point(0, y_offset));
```
Comment (jonathanlew) [inline src/software/ai/hl/stp/play/indirect_free_kick_play.cpp]: I think divide by double would look better
```suggestion
            Rectangle(Point(0, world.field().xLength() / 4.0), Point(0, y_offset));
```

File: src/software/ai/hl/stp/play/indirect_free_kick_play.cpp [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+        std::make_shared<CherryPickTactic>(world, cherry_pick_2_target_region);
+
+    // This tactic will move a robot into position to initially take the free-kick
+    auto align_to_ball_tactic = std::make_shared<MoveTactic>();
+    align_to_ball_tactic->addWhitelistedAvoidArea(AvoidArea::BALL);
+
+    PassGenerator pass_generator(world, world.ball().position(),
+                                 PassType::RECEIVE_AND_DRIBBLE);
+    pass_generator.setTargetRegion(
+        Rectangle(Point(0, world.field().yLength() / 2), world.field().enemyCornerNeg()));
```
Comment (jonathanlew) [inline src/software/ai/hl/stp/play/indirect_free_kick_play.cpp:107]: why that corner?

**Review by jonathanlew — CHANGES_REQUESTED** [review]
Just some comments to address...

**Review by akhilveeraghanta — DISMISSED** [review]
Looks Good :tm:

File: src/software/ai/hl/stp/play/corner_kick_play.h (hunk 1 above)
Comment (MathewMacDougall) [reply src/software/ai/hl/stp/play/corner_kick_play.h:29]: The indirect free kick play needed access to the `BALL_IN_CORNER_RADIUS`

File: src/software/ai/hl/stp/play/indirect_free_kick_play.cpp (hunk 3 above)
Comment (MathewMacDougall) [reply src/software/ai/hl/stp/play/indirect_free_kick_play.cpp:107]: This is actually the entire enemy half of the field, not just a corner. Maybe we want member function to get the middle sideline points of the field? Not sure the right name, but can make a ticket if we think this is something we use relatively often (which I think we do).

**Review by jonathanlew — DISMISSED** [review]
:+1:

File: src/software/ai/hl/stp/play/indirect_free_kick_play.cpp (hunk 3 above)
Comment (garethellis0) [reply src/software/ai/hl/stp/play/indirect_free_kick_play.cpp:107]: maybe it's worth adding `getPositiveXHalf()` and `getNegativeXHalf()` functions to the `Field` class?

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Ok, suggested a few places where there are _really_ obvious and simple ways to pull stuff into functions. Refrained from doing too many since we're doing refactors of all these later.

File: src/software/ai/hl/stp/play/indirect_free_kick_play.cpp (hunk 3 above)
Comment (MathewMacDougall) [reply src/software/ai/hl/stp/play/indirect_free_kick_play.cpp:107]: created #979

File: src/software/ai/hl/stp/play/indirect_free_kick_play.cpp [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+
+bool IndirectFreeKickPlay::isApplicable(const World &world) const
+{
+    double min_dist_to_corner =
+        std::min((world.field().enemyCornerPos() - world.ball().position()).len(),
+                 (world.field().enemyCornerNeg() - world.ball().position()).len());
+
+    // Make sure we don't interfere with the cornerkick play
+    return world.gameState().isOurFreeKick() &&
+           min_dist_to_corner >= CornerKickPlay::BALL_IN_CORNER_RADIUS;
```
[the remark this answers is not in the document]
Comment (MathewMacDougall) [reply src/software/ai/hl/stp/play/indirect_free_kick_play.cpp:39]: Can we do this as part of refactoring? That seems out of scope for this PR, but not a bad thing to revisit. Added to #968

[the remark this answers is not in the document]
Comment (MathewMacDougall) [reply src/software/ai/hl/stp/play/indirect_free_kick_play.cpp:39]: Actually I don't think they should be made static because then we lose the benefit of inheritance / enforced interfaces, but that's a discussion for another time.

**Comment by garethellis0** [comment]
@MathewMacDougall - I originally wrote this, so lemme know if you want me to to some of the refactoring comments and I can open a PR to your branch

**Comment by MathewMacDougall** [comment]
Taking a small break to work on my documentation tickets, will hopefully update this by Thursday.

@garethellis0 you're pretty busy so I'll take a shot at your suggested changes.

File: src/software/ai/hl/stp/play/indirect_free_kick_play.cpp [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+            chip_tactic->updateWorldParams(world.ball());
+            chip_tactic->updateControlParams(world.ball().position(), chip_target,
+                                             chip_dist);
+            goalie_tactic->updateWorldParams(world.ball(), world.field(),
+                                             world.friendlyTeam(), world.enemyTeam());
+
+            yield({goalie_tactic, chip_tactic, std::get<0>(crease_defender_tactics),
+                   std::get<1>(crease_defender_tactics)});
+
+        } while (!chip_tactic->done());
```
[the remark this answers is not in the document]
Comment (MathewMacDougall) [reply src/software/ai/hl/stp/play/indirect_free_kick_play.cpp]: done

File: src/software/ai/hl/stp/play/indirect_free_kick_play.cpp [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+            receiver->updateWorldParams(world.friendlyTeam(), world.enemyTeam(),
+                                        world.ball());
+            receiver->updateControlParams(pass);
+            receiver->addWhitelistedAvoidArea(AvoidArea::BALL);
+            goalie_tactic->updateWorldParams(world.ball(), world.field(),
+                                             world.friendlyTeam(), world.enemyTeam());
+
+            yield({goalie_tactic, passer, receiver, std::get<0>(crease_defender_tactics),
+                   std::get<1>(crease_defender_tactics)});
+        } while (!receiver->done());
```
[the remark this answers is not in the document]
Comment (MathewMacDougall) [reply src/software/ai/hl/stp/play/indirect_free_kick_play.cpp]: done. Leaving the crease defender and goalie as local variables passed to the function for now.

**Comment by MathewMacDougall** [comment]
Changes mostly done will push in a bit over an hour, just commenting as I go.

File: src/software/ai/hl/stp/play/indirect_free_kick_play.cpp [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+        best_pass_and_score_so_far = pass_generator.getBestPassSoFar();
+        LOG(DEBUG) << "Best pass found so far is: " << best_pass_and_score_so_far.pass;
+        LOG(DEBUG) << "    with score: " << best_pass_and_score_so_far.rating;
+
+        Duration time_since_commit_stage_start =
+            world.getMostRecentTimestamp() - commit_stage_start_time;
+        min_score = 1 - std::min(time_since_commit_stage_start.getSeconds() /
+                                     MAX_TIME_TO_COMMIT_TO_PASS.getSeconds(),
+                                 1.0);
+    } while (best_pass_and_score_so_far.rating < min_score);
```
[the remark this answers is not in the document]
Comment (MathewMacDougall) [reply src/software/ai/hl/stp/play/indirect_free_kick_play.cpp]: done

**Comment by MathewMacDougall** [comment]
addressed PR comments and fixed merge conflict. Ready for re-review.

**Review by jonathanlew — DISMISSED** [review]
:+1:

**Review by garethellis0 — DISMISSED** [review]
LGTM :+1: :tada: - nice job breaking stuff into functions, looks much cleaner.

**Comment by MathewMacDougall** [comment]
Fixed the build, please re-approve.

**Review by jonathanlew — APPROVED** [review]

**Review by akhilveeraghanta — APPROVED** [review]
Looks Good :tm:

**Review by garethellis0 — APPROVED** [review]
LGTM :+1: :tada:
